### DISCUSSION TITLE
Adding a JSON based channel for symbology specific config/result info?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#862] [Polls]
There are a couple of recent request for change or issues that asked for a change in the API:
 * export QRCode Mask (#849)
 * allow to force a QRCode Version during encoding (#861)
 * allow to lower the minimum length for ITF symbols (#853)

All of them have in common that they are
 1. pretty niche in their application
 2. specific to only one symbology
 3. need an API change

Given that there are so many wrappers now that need to be kept in sync, which incurs a substantial maintenance cost, there needs to be a pretty high value added for justifying this, in my opinion. The simple addition of a new `BarcodeFormat::DataBarLimited` constant, that I did recently, already took me a substantial chunk of time. None of the 3 mentioned issues would cross that threshold for me.

As mentioned [here](https://github.com/zxing-cpp/zxing-cpp/pull/849#issuecomment-2411243680), this reminded me of the meta data key-value map type data structure in the original Java API that was solving basically this use case. I didn't like that complex API design, though. So my idea would be to use the simplest and most generic channel to communicate this kind of data: a JSON string. Once established, it would completely eliminate any work across the wrappers when adding yet another bit of info. On the down-side, it would require the client code do deal with JSON if they are interested in the "extra" information and could lead to a bifurcation of the API since once we have that, why then not add all the existing bits of data to that JSON string as well?

**EDIT**: With the release of 3.0 this new API is now official and stable. For details see
* `Barcode::extra()` and `namespace BarcodeExtra`
* `CreatorOptions::options`

**Comment by markusfisch** [comment]
Perhaps a `std::map<std:: string, std::any>` would also be sufficient. Presumably this information never needs to be nested, a map/dictionary is available in all languages and would save us the JSON parsing. On the other hand, this would of course be a one-off effort in every wrapper.

**Reply by axxel** [reply]
The `std::any` part sounds a bit 'scary' to me. First of all, I don't now how that would be modeled in e.g. kotlin/native. And even if we check that we have something appropriate in every current wrapper, I can imagine that there might come along a new one where that non-typed any-value would be an issue. Second of all: every wrapper needs to know the type of the individual value. A `std::string` does not have either of these issues.

Also what is the canonical dictionary data structure in C? I would argue that serializing the data into a string would be the easiest way to transport the map from c++ via c to .NET, etc. But each wrapper (that cares about it) could add a trivial "flat JSON to dictionary" function to not have all client-code need to do that?

**Comment by synedra-mpe** [comment]
Also #869 might be a good candidate for that JSON based channel.

EDIT (by axxel): Just for quicker parsing and centralizing the discussion, here is the gist of this proposal:

Add information about the **quiet zone**, or lack thereof. Something along the lines of
```JSON
"endQuietZoneDetected": "true",
"fullIsoCompliant": "true",
"endQuietZoneLength": "10" or
"cropped": "true"
```

**Reply by axxel** [reply]
A first draft of the JSON channel is implemented but I failed to quickly add this feature here since it is not obvious to me what exactly is of interest here:
1. the quiet zone in pixels/modules? That would not be very helpful to the general user as they would not know how to interpret the number.
2. a plain `hasIsoCompliantQuietZone` boolean? That would not really have helped the issue in #869 as often the quiet zones are not compliant but still work, e.g. ITF requires 10 modules, while `zxing-cpp` is fine with 6 to not loose too many real-world examples.
3. something like `croppedQuietZone`? Is that true if one side is cropped or only if both are?

I'd like make sure the data we add is actually useful and as generic (across symbologies) as possible. Suggestions?

**Reply by synedra-mpe** [reply]
ITF detection depends heavily on how quiet zones are handled by the scanner, at least when scanning with a live camera, as can be seen in #869.
Something like `croppedQuietZone` would have been helpful in this regard, as it would have been possible to ignore all detection results without iso-compliant quiet zones.

**Comment by axxel** [comment]
Another application might be to provide information about whether or not the **encoding has been guessed**. Obviously @marcoffee saw a [need for this](https://github.com/marcoffee/zxing-cpp/commit/3c69ecac0d954a93d43995a74a9b490755e53eaf).

**Reply by axxel** [reply]
@marcoffee Would you rather have the option to completely suppress the use of the guesswork, something like `enforceIsoCompliantDecoding` in the reader options? Or are you more interested in detecting the cases where the symbols are (likely) not ISO compliant, namely those where the library currently engages in guesswork?

**Reply by marcoffee** [reply]
@axxel I'm more interested in detecting those cases because the guessed symbols might be what I need (e.g., if they pass a regex check or something like that). I defined the method just to have a way to identify if it requires further validation.

In my opinion, the other way (suppressing guesswork), has two issues:
1) if we suppress, we lose information that there is a barcode in the image (which may lead to information/processing loss)
2) if we don't suppress, we will have no way of knowing if the text is correct (which may lead to runtime errors)

So, I think that having this flag is a good compromise between having the information but making the choice to either discard it completely or use/fix it.

**Reply by axxel** [reply]
> 1. if we suppress, we lose information that there is a barcode in the image

That is not correct. What I meant is to instruct `zxing-cpp` to strictly follow the specification while decoding the bytes into a utf8 string. That means, if there is no explicit information about the used encoding, then the default encoding of the symbology (usually ISO8859_1) is to be used. In an ideal world, this would be the only sensible approach. Unfortunately, there are lots of symbols out there, especially QRCodes, that violate the specification by encoding e.g. SHIFT_JIS (Japanese) text but fail to set the proper ECI metadata. Those bytes come out as garbage then, if interpreted as ISO8859_1 (as required by the spec). See e.g. [this recent bug in Java ZXing](https://github.com/zxing/zxing/issues/1500). The QRCode situation is especially contrived due to the additional complexity introduced by the compaction modes. If you have time to waste, have a look at the discussion before and after [this comment](https://github.com/zxing/zxing/pull/1498#issuecomment-1134578616). Reading this again myself made me aware that the current implementation actually has an almost trivial check for whether or not there was guess-work in action: if the symbology supports ECI encoding (QRCode, Aztec, Maxicode, PDF417, DataMatrix) and `Barcode::hasECI() == false`, then there is guessing involved.

**Reply by marcoffee** [reply]
Hmmm. I see.
I thought  it would completely ignore the barcode if the specification could not be detected.

In that case, I like the idea of having both options. A flag to completely disable guessing along with a parameter indicating whether the encoding was guessed (`true` if the flag was enabled and it was guessed) or not (`false` if the flag was not enabled or the encoding is exact).

That way, we can have the expertise of `zxing-cpp` to guess the encoding and a way to detect if we need to post-process that decoding. In addition, a list of detected encodings (with their start/end ranges on the data) would be interesting for the post-processing step. Imagine a case where you know that a barcode only come with a handful of encodings. Use that list, you can tell for sure if the guessed encoding is wrong. This would also be useful as an input.

Maybe we can join both ideas and allow passing this list of 'guessable' encodings instead of a flag, that way, we can disallow guesses using an empty list or allow any kind of guessing by passing all available encodings (which would be a constant).

**Reply by axxel** [reply]
It seems a couple of things need clarification:
1. The only encodings that could potentially be guessed are `ISO_88591`, `UTF-8` and `SHIFT-JIS`. This is mostly geared toward the QRCode use case.
2. The current code already allows you to determine if guessing was involved without any extra API: `bool isGuessed(const Barcode& b) { return !b.hasECI() && (QRCode | Aztec | Maxicode | PDF417 | DataMatrix).testFlag(b.format()); }`
3. `Barcode::bytes()` returns the 'raw' data without any guessing or code page conversion. If you know the encoding, simply apply it to that stream of bytes.
4. I'm considering changing the definition of the `TextMode::ECI` flag to implicitly mean "interpret encoding according to standard". This would then allow you to get hold of that data without any guessing and without any extra API. You would only need to ignore the ECI designators and undo the doubling of `\`.

**Reply by axxel** [reply]
@gitlost and @oehhar do you know from the top of your head if there is any symbology currently supported that has a default encoding other than ISO_8859-1?

**Reply by oehhar** [reply]
Yes, JAB-Code has utf-8. But it is not supported by zxing-cpp ;-).

When you write, that you want to change the output of TextMode::ECI:

For me, one mode would be great: do what a standard scanner does:
if (hasECI) return bytesECI else return symbologyIdentifier+bytes

**Reply by axxel** [reply]
> For me, one mode would be great: do what a standard scanner does:
> if (hasECI) return bytesECI else return symbologyIdentifier+bytes

That is _exactly_ what `Barcode::bytesECI()` does (or is supposed to do, anyway).

**Reply by oehhar** [reply]
Thanks, great !
Lets look to two examples:

The first code is a system login code with non ISO-8859-1 characters, e.g. containing an ECI:

![code_eci](https://github.com/user-attachments/assets/b4328fcc-ed81-4885-a758-3544b926724b)
The code contains an ECI 2 for ISO-Latin-2 and the data "1HJ" then a byte with decimal value 234 then "drek".
The byte with decimal 234 in ISO8859-2 is unicode codepoint x0119 "ę"
The decode result is as follows. Remark that byte 234 is shown by its Unicode representation  "ê".
- text 1HJędrek
- format DataMatrix
- bytes 1HJêdrek
- bytesECI ]d4\0000041HJêdrek
- content Text
- symbologyIdentifier \]d1
- hasECI 1

Here, bytesECI is, what a normal scanner would output. The item "text" with the decoded ECI is great (remark the change of "ê" to "ę" due to the ISO-8859-2 decoding).

The output of a standard scanner is exactly what "bytesECI" contains.

Now try a HIBC code without any ECI which only contains printable ASCII:

![code_no_eci](https://github.com/user-attachments/assets/c808b048-c469-486f-8b93-7b6d656a5691)

The decode result is:

- text +E4947611201/$LI
- format DataMatrix
- bytes +E4947611201/$LI
- bytesECI ]d4+E4947611201/$LI
- content Text
- symbologyIdentifier \]d1
- hasECI 0

The output of a standard scanner is what the concatenation of "symbologyIdentifier" and "bytes" contains.

To get the standard output in both cases, I do:

if (hasECI) return bytesECI else return symbologyIdentifier+bytes

Thanks for all the great work !
Harald

**Reply by axxel** [reply]
Thanks a lot for this insight. Let's follow up this discussion in #936.

**Comment by axxel** [comment]
There is some progress to be reported. I added a first [draft](https://github.com/zxing-cpp/zxing-cpp/commit/f67271fc618e296c142ac38e6ce1808ab63417a9) that provides access to the `DataMask` and proper `Version` of a QRCode as well as the original/unexpanded value of a `UPC-E` symbol.

There are many open questions:
1. What to call the property of the `Barcode` class that holds that JSON string? My current 2 best options would be `properties` or `extra`. The former would lend itself well in case we decide to actually provide all properties as a JSON object, while it does expressly not if we don't :-/. Other ideas?
EDIT: Obviously `meta` would be another contender, if only that term wasn't so 'tainted' :-/.
2. What name scheme should we use for the key names: `dataMask` or `DataMask`? The first is more in line with the names of the other `Barcode` class members, the latter would look better if the JSON is simply printed on the screen as is.
3. Should the key names give an idea where they are coming from: `DataMask` vs. `QRCodeDataMask`? I slightly prefer the former, as the symbology is already specified elsewhere.

A WIP draft is also available for the [CreatorOptions](https://github.com/zxing-cpp/zxing-cpp/commit/f825e96550da9e12c3c6f800744915997274e7a6#diff-f0ed128f78a05c1a0ff56be8928966ee3eee8237ce6e2f56f936e4bb1513d189) use case.

Comments welcome.

**Reply by axxel** [reply]
[This patch](https://github.com/zxing-cpp/zxing-cpp/commit/17ffc9c7e80e5c70ab0479dc77f023eb192fb61f) adds another usability improvement to the idea, completely removing the need to parse a JSON string in case one knows the key name in advance. Noteworthy: key-names are compared case insensitive.

**Reply by axxel** [reply]
Current HEAD includes [this addition](https://github.com/zxing-cpp/zxing-cpp/commit/83d6ef2e6c01e3991f78ad861e146325870cd4fd) to `ZXingReader` that allows to output a (more or less) complete JSON serialization of the barcode object, including JSON conform [character escaping](https://github.com/zxing-cpp/zxing-cpp/commit/cfdf0d5745f509c629a9ab2fa7cd066b925ab409) which is something I overlooked when first drafting this idea.

**Reply by axxel** [reply]
With the release of 3.0, this matter has largely been settled. I'm letting this disussion open, in case someone wants to add ideas about which information should be added to the new JSON channel(s).

**Comment by gitlost** [comment]
Of those supported by Zint only Grid Matrix (default GB 2312) and
UPNQR (default ISO 8859-2), a QR variant specific to Slovenian banks.

**Comment by isning** [comment]
Just here to provide another choice:
I wonder if anyone has considered that using Protobuf definitions to generate class definitions in wrappers might be a win-win approach. It can be convenient for client so that they don't need to deal with json parsing and ideally, there should be no much effort need to deal with wrappers when api changes happens.

There's also some disadvantage:
- package size may increase since there'll be a protobuf parser or something be bundled
- much effort needed to add support for this for wrappers
- maybe not all wrapper "natively" hav out-of-box "class generator" solution

**Reply by axxel** [reply]
Thanks for sharing your idea but I don't see myself investing time in that direction. Without even looking at the details, it sounds like a lot of work, which was exactly what I was trying to avoid in the first place. Also my gut feeling tells me that this would go against my goal of having all wrappers feel native to their language and not be forced into some common denominator API.
